flyway-community-db-support/src/main/java/org/flywaydb/community/database/clickhouse/ClickHouseDatabase.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (C) Red Gate Software Ltd 2010-2022
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.flywaydb.community.database.clickhouse;
+
+import org.flywaydb.core.api.configuration.Configuration;
+import org.flywaydb.core.internal.database.base.Database;
+import org.flywaydb.core.internal.database.base.Table;
+import org.flywaydb.core.internal.jdbc.JdbcConnectionFactory;
+import org.flywaydb.core.internal.jdbc.StatementInterceptor;
+import org.flywaydb.core.internal.util.StringUtils;
+
+import java.sql.Connection;
+
+public class ClickHouseDatabase extends Database<ClickHouseConnection> {
+    @Override
+    public boolean useSingleConnection() {
+        return true;
+    }
+
+    public ClickHouseDatabase(Configuration configuration, JdbcConnectionFactory jdbcConnectionFactory, StatementInterceptor statementInterceptor) {
+        super(configuration, jdbcConnectionFactory, statementInterceptor);
+    }
+
+    public String getClusterName() {
+        return configuration.getPluginRegister().getPlugin(ClickHouseConfigurationExtension.class).getClusterName();
+    }
+
+    @Override
+    protected ClickHouseConnection doGetConnection(Connection connection) {
+        return new ClickHouseConnection(this, connection);
+    }
+
+    @Override
+    public void ensureSupported() {
+    }
+
+    @Override
+    public boolean supportsDdlTransactions() {
+        return false;
+    }
+
+    @Override
+    public boolean supportsMultiStatementTransactions() {
+        return false;
+    }
+
+    @Override
+    public String getBooleanTrue() {
+        return "1";
+    }
+
+    @Override
+    public String getBooleanFalse() {
+        return "0";
+    }
+
+    @Override
+    public boolean catalogIsSchema() {
+        return true;
+    }
+
+    @Override
+    public String getRawCreateScript(Table table, boolean baseline) {
+        String clusterName = getClusterName();
+        boolean isClustered = StringUtils.hasText(clusterName);
+
+        String script = "CREATE TABLE IF NOT EXISTS " + table + (isClustered ? (" ON CLUSTER " + clusterName) : "") + "(" +
+                "    installed_rank Int32," +
+                "    version String," +
```

Comment:
Unfortunately, on repeatable migrations version will be null and that failed with:
```
ERROR: Failed to execute operation in non-transactional context. Please restore backups and roll back database and code!
ERROR: Unable to insert row for version 'null' in Schema History table "datamart"."__flyway_schema_history"
----------------------------------------------------------------------------------------------------
SQL State  : HY000
Error Code : 0
Message    : Cannot set null to non-nullable column #2 [version String]

Caused by: java.sql.SQLException: Cannot set null to non-nullable column #2 [version String]
```

`installed_rank` should be PRIMARY KEY and not-nullable, but version `Nullable(String)`.

So, resulting table should look like:
```sql
CREATE TABLE datamart.__flyway_schema_history_NEW ON CLUSTER ${CLUSTER_NAME} (
    `installed_rank` Int32,
    `version` Nullable(String),
    `description` String,
    `type` String,
    `script` String,
    `checksum` Nullable(Int32),
    `installed_by` String,
    `installed_on` DateTime DEFAULT now(),
    `execution_time` Int32,
    `success` Bool
)
ENGINE = MergeTree
PRIMARY KEY `installed_rank`
ORDER BY `installed_rank`
SETTINGS index_granularity = 8192;
```

Reply:
I think it's better to set field 'script' as primary key. Field 'installed_rank has no connection with migration content.

Reply:
@Hubbitus could you please check the fix.

Reply:
For postgres variant `installed_rank` is the primary key.

Reply:
Maybe they should change that for PG as well.

Reply:
[Ignite as well](https://github.com/flyway/flyway/blob/08ff5b0c6ecf2a87e32e41d72904b7ec92a522de/flyway-community-db-support/src/main/java/org/flywaydb/community/database/ignite/thin/IgniteThinDatabase.java#L75), I also checked `mysql`, `sqlite` and some others. All has `installed_rank` as `PRIMARY KEY`.
Meantime, it looks like for the Clickhouse we may omit it completely? So, there is even no unique constraint will be implied...

I think it is not consistent to set it to the `script` value. But I'm uncertain.

Reply:
Primary key in this table prevents duplications when applying migrations, but field 'installed_rank' contains auto incremented value and it's useless for this purpose. Field 'script' contains version number and migration name, it's unique and it's a perfect candidate for the primary key. Let's wait for maintainer to review this PR.